File: CS/GrowthWare.WebSupport/Utilities/MenuUtility.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GrowthWare.WebSupport.Utilities
{
    /// <summary>
    /// Class MenuUtility
    /// </summary>
    public static class MenuUtility
    {
        /// <summary>
        /// Generates and order list form hierarchical data.
        /// </summary>
        /// <param name="menuData">The menu data.</param>
        /// <param name="value">The string builder.</param>
        /// <returns>String.</returns>
        public static String GenerateUnorderedList(DataTable menuData, StringBuilder value)
        {
            if (value == null) throw new ArgumentNullException("value", "value cannot be a null reference (Nothing in Visual Basic)!");
            if (menuData == null) throw new ArgumentNullException("menuData", "menuData cannot be a null reference (Nothing in Visual Basic)!");
            value.AppendLine("<ul>");
            DataView datView = null;
            try
            {
                datView = new DataView(menuData);
                datView.RowFilter = "ParentID = 1";
                //Populate menu with top menu items;
                foreach (DataRowView rowVeiw in datView)
                {
                    //Define new menu item
                    if (int.Parse(rowVeiw["FUNCTION_TYPE_SEQ_ID"].ToString(), CultureInfo.InvariantCulture) == 3)
                    {
                        value.AppendLine(createLIItem(rowVeiw["Title"].ToString(), rowVeiw["URL"].ToString(), rowVeiw["Description"].ToString(), true));
                    }
                    else
                    {
                        value.AppendLine(createLIItem(rowVeiw["Title"].ToString(), rowVeiw["URL"].ToString(), rowVeiw["Description"].ToString(), false));
                    }
                    //Populate child items of this parent
                    addChildItems(menuData, int.Parse(rowVeiw["MenuID"].ToString(), CultureInfo.InvariantCulture), ref value);
                }
            }
            catch (Exception)
            {
                throw;
            }
            finally 
            {
                if (datView != null) datView.Dispose();
            }

            value.AppendLine("<ul>");
            return value.ToString();
        }

        /// <summary>
        /// Add the child items to the stringBuilder
        /// </summary>
        /// <param name="menuData">DataTable</param>
        /// <param name="parentID">Integer</param>
        /// <param name="stringBuilder">stringBuilder</param>
        /// <remarks></remarks>
        private static void addChildItems(DataTable menuData, int parentID, ref StringBuilder stringBuilder)
        {
            //Populate DataView
            DataView datView = null;
            try
            {
                datView = new DataView(menuData);
                //Filter child menu items
                datView.RowFilter = "parentid = " + parentID;
                //Populate parent menu item with child menu items
                stringBuilder.AppendLine("<ul>");
                foreach (DataRowView datRow in datView)
                {
                    //Define new menu item
                    if (int.Parse(datRow["FUNCTION_TYPE_SEQ_ID"].ToString(), CultureInfo.InvariantCulture) == 3)
                    {
                        stringBuilder.AppendLine(createLIItem(datRow["Title"].ToString(), datRow["URL"].ToString(), datRow["Description"].ToString(), true));
                    }
                    else
                    {
                        stringBuilder.AppendLine(createLIItem(datRow["Title"].ToString(), datRow["URL"].ToString(), datRow["Description"].ToString(), false));
                    }
                    //stringBuilder.AppendLine(createLIItem(datRow("Title"), datRow("URL"), datRow("Description"), mHasChildren))
                    //Populate child items of this parent
                    addChildItems(menuData, int.Parse(datRow["MenuID"].ToString(), CultureInfo.InvariantCulture), ref stringBuilder);
                }
            }
            catch (Exception)
            {
                throw;
            }
            finally 
            {
                if (datView != null) datView.Dispose();
            }
            stringBuilder.AppendLine("</ul>");
        }

        /// <summary>
        /// Add the li with either the class has-sub or without
        /// </summary>
        /// <param name="hrefText">Link Text</param>
        /// <param name="action">Used to build the javascript</param>
        /// <param name="hrefToolTip">Used for the tooltip property</param>
        /// <param name="hasChildren">Determines the class='has-sub' for the li tag</param>
        /// <returns></returns>
        /// <remarks></remarks>
        private static String createLIItem(String hrefText, String action, String hrefToolTip, Boolean hasChildren)
        {
            String retVal = String.Empty;
            if (!hasChildren)
            {
                retVal = "<li><a href=\"" + action + "\" title=\"" + hrefToolTip + "\"><span>" + hrefText + "</span></a>";
            }
            else
            {
                retVal = "<li class='has-sub'><a href='#' title=\"" + hrefToolTip + "\"><span>" + hrefText + "</span></a>";
            }
            return retVal;
        }
    }
}

File: VB/GrowthWare.WebSupport/Utilities/MenuUtility.vb
Imports System.Text
Imports System.Globalization

Namespace Utilities
    ''' <summary>
    ''' Utiltiy class to aid in a web application
    ''' </summary>
    ''' <remarks>Could be considered Specific to Growthware</remarks>
    Public Module MenuUtility

        ''' <summary>
        ''' Generates and order list form hierarchical data.
        ''' </summary>
        ''' <param name="menuData">Hierarchical datatable</param>
        ''' <param name="value">StringBuiler used to build the ul/li string data</param>
        ''' <returns>String</returns>
        ''' <remarks>Frist Layer of items should have a ParentID of 1</remarks>
        Public Function GenerateUnorderedList(ByVal menuData As DataTable, ByVal value As StringBuilder) As String
            If value Is Nothing Then Throw New ArgumentNullException("value", "value cannot be a null reference (Nothing in Visual Basic)!")
            value.AppendLine("<ul>")
            Dim datView As DataView = Nothing
            Try
                datView = New DataView(menuData)
                datView.RowFilter = "ParentID = 1"
                '//Populate menu with top menu items
                Dim datRow As DataRowView
                For Each datRow In datView
                    '//Define new menu item
                    If Integer.Parse(datRow("FUNCTION_TYPE_SEQ_ID").ToString(), CultureInfo.InvariantCulture) = 3 Then
                        value.AppendLine(createLIItem(datRow("Title"), datRow("URL"), datRow("Description"), True))
                    Else
                        value.AppendLine(createLIItem(datRow("Title"), datRow("URL"), datRow("Description"), False))
                    End If
                    '//Populate child items of this parent
                    addChildItems(menuData, datRow("MenuID"), value)
                Next
            Catch ex As Exception
                Throw
            Finally
                If Not datView Is Nothing Then datView.Dispose()
            End Try
            value.AppendLine("<ul>")
            Return value.ToString()
        End Function

        ''' <summary>
        ''' Add the child itmes to the StringBuiler
        ''' </summary>
        ''' <param name="menuData">DataTable</param>
        ''' <param name="parentID">Integer</param>
        ''' <param name="stringBuilder">StringBuiler</param>
        ''' <remarks></remarks>
        Private Sub addChildItems(ByVal menuData As DataTable, ByVal parentID As Integer, ByRef stringBuilder As StringBuilder)
            '//Populate DataView
            Dim datView As DataView = Nothing
            Try
                datView = New DataView(menuData)
                '//Filter child menu items
                datView.RowFilter = "parentid = " & parentID
                '//Populate parent menu item with child menu items
                Dim datRow As DataRowView
                stringBuilder.AppendLine("<ul>")
                For Each datRow In datView
                    '//Define new menu item
                    If Integer.Parse(datRow("FUNCTION_TYPE_SEQ_ID").ToString(), CultureInfo.InvariantCulture) = 3 Then
                        stringBuilder.AppendLine(createLIItem(datRow("Title"), datRow("URL"), datRow("Description"), True))
                    Else
                        stringBuilder.AppendLine(createLIItem(datRow("Title"), datRow("URL"), datRow("Description"), False))
                    End If

                    'stringBuilder.AppendLine(createLIItem(datRow("Title"), datRow("URL"), datRow("Description"), mHasChildren))
                    '//Populate child items of this parent
                    addChildItems(menuData, datRow("MenuID"), stringBuilder)
                Next
            Catch ex As Exception
                Throw
            Finally
                If Not datView Is Nothing Then datView.Dispose()
            End Try
            stringBuilder.AppendLine("</ul>")
        End Sub

        ''' <summary>
        ''' Add the li with either the class has-sub or without
        ''' </summary>
        ''' <param name="hrefText">Link Text</param>
        ''' <param name="action">Used to build the javascript</param>
        ''' <param name="hrefToolTip">Used for the tooltip property</param>
        ''' <param name="hasChildren">Determines the class='has-sub' for the li tag</param>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Private Function createLIItem(ByVal hrefText As String, ByVal action As String, ByVal hrefToolTip As String, ByVal hasChildren As Boolean) As String
            Dim retVal As String = String.Empty
            If Not hasChildren Then
                retVal = "<li><a href=""" + action + """ title=""" + hrefToolTip + """><span>" + hrefText + "</span></a>"
            Else
                retVal = "<li class='has-sub'><a href='#' title=""" + hrefToolTip + """><span>" + hrefText + "</span></a>"
            End If
            Return retVal
        End Function
    End Module
End Namespace
